src/libmeasurement_kit/dns/system_resolver.hpp

```diff
@@ -0,0 +1,174 @@
+            break;
+        case EAI_BADHINTS:
+            resolver_error = InvalidHintsValue();
+            break;
+        case EAI_FAIL:
+            resolver_error = NonRecoverableFailure();
+            break;
+        case EAI_FAMILY:
+            resolver_error = NotSupportedAIFamily();
+            break;
+        case EAI_MEMORY:
+            resolver_error = MemoryAllocationFailure();
+            break;
+        case EAI_NONAME:
+            resolver_error = HostOrServiceNotProvidedOrNotKnown();
+            break;
+        case EAI_OVERFLOW:
+            resolver_error = ArgumentBufferOverflow();
+            break;
+        case EAI_PROTOCOL:
+            resolver_error = UnknownResolvedProtocol();
+            break;
+        case EAI_SERVICE:
+            resolver_error = NotSupportedServname();
+            break;
+        case EAI_SOCKTYPE:
+            resolver_error = NotSupportedAISocktype();
+            break;
+        default:
+            resolver_error = ResolverError();
+        }
+        ctx->reactor->call_soon([=]() { callback(resolver_error, nullptr); });
+        return;
+    }
+    assert(servinfo != nullptr);
+
+    void *addr_ptr;
+    char address[128];
+    for (addrinfo *p = servinfo.get(); p != nullptr; p = p->ai_next) {
+        Answer answer;
+        answer.name = ctx->name;
+        answer.qclass = ctx->dns_class;
+        assert(p->ai_family == AF_INET or p->ai_family == AF_INET6);
+        if (p->ai_family == AF_INET) {
+            answer.type = QueryTypeId::A;
+            addr_ptr = &((sockaddr_in *)p->ai_addr)->sin_addr;
+        } else if (p->ai_family == AF_INET6) {
+            answer.type = QueryTypeId::AAAA;
+            addr_ptr = &((sockaddr_in6 *)p->ai_addr)->sin6_addr;
+        }
+        if (p->ai_canonname != nullptr) {
+            answer.hostname = p->ai_canonname;
+        }
+        if (inet_ntop(p->ai_family, addr_ptr, address, sizeof(address)) ==
+            nullptr) {
+            ctx->logger->warn("dns: unexpected inet_ntop failure");
+            ctx->reactor->call_soon(
+                [=]() { callback(InetNtopFailure(), nullptr); });
+            return;
+        }
+        if (p->ai_family == AF_INET) {
+            answer.ipv4 = std::string(address);
+        } else if (p->ai_family == AF_INET6) {
+            answer.ipv6 = std::string(address);
+        }
+        message->answers.push_back(answer);
+    }
+    ctx->reactor->call_soon([=]() { callback(NoError(), message); });
+}
+
+template <MK_MOCK(getaddrinfo), MK_MOCK(inet_ntop)>
+void system_resolver(QueryClass dns_class, QueryType dns_type, std::string name,
+                     Callback<Error, Var<Message>> cb, Settings settings,
+                     Var<Reactor> reactor, Var<Logger> logger) {
+    std::unique_ptr<ResolverContext> ctx(new ResolverContext(
+        dns_class, dns_type, name, cb, settings, reactor, logger));
+    Query query;
+    ctx->hints.ai_flags = AI_ALL | AI_V4MAPPED;
```

Comment:
It should return (used with AI_ALL) the ipv6 address and the ipv4 to ipv6 mapped addresses when we ask for AAAA resolution.